db/sovereign-db/src/ledger_db.rs

```diff
@@ -0,0 +1,245 @@
+            "Number of transactions must match number of event groupss."
+        );
+        Ok(commit)
+    }
+}
+
+pub struct SlotCommit {
+    pub slot_data: StoredSlot,
+    pub batches: Vec<StoredBatch>,
+    pub txs: Vec<StoredTransaction>,
+    pub events: Vec<Vec<Event>>,
+}
+
+impl SlotCommitBuilder {}
+
+impl<S: SlotData> LedgerDB<S> {
+    pub fn with_path(path: impl AsRef<Path>) -> Result<Self, anyhow::Error> {
+        let path = path.as_ref().join(LEDGER_DB_PATH_SUFFIX);
+        let inner = DB::open(
+            path,
+            "ledger-db",
+            LEDGER_TABLES.iter().copied(),
+            &gen_rocksdb_options(&Default::default(), false),
+        )?;
+
+        let next_item_numbers = ItemNumbers {
+            slot_number: Self::last_version_written(&inner, SlotByNumber)?.unwrap_or_default() + 1,
+            batch_number: Self::last_version_written(&inner, BatchByNumber)?.unwrap_or_default()
+                + 1,
+            tx_number: Self::last_version_written(&inner, TxByNumber)?.unwrap_or_default() + 1,
+            event_number: Self::last_version_written(&inner, EventByNumber)?.unwrap_or_default()
+                + 1,
+        };
+
+        Ok(Self {
+            db: Arc::new(inner),
+            slots_to_execute: Default::default(),
+            next_item_numbers: Arc::new(Mutex::new(next_item_numbers)),
+        })
+    }
+
+    /// A rocksdb instance which stores its data in a tempdir
+    #[cfg(any(test, feature = "temp"))]
+    pub fn temporary() -> Self {
+        let path = schemadb::temppath::TempPath::new();
+        Self::with_path(path).unwrap()
+    }
+
+    pub fn get_next_items_numbers(&self) -> ItemNumbers {
+        self.next_item_numbers.lock().unwrap().clone()
+    }
+
+    fn put_slot(&self, slot: &StoredSlot, slot_number: &SlotNumber) -> Result<(), anyhow::Error> {
+        self.db.put::<SlotByNumber>(slot_number, slot)?;
+        self.db.put::<SlotByHash>(&slot.hash, slot_number)
+    }
+
+    fn put_batch(
+        &self,
+        batch: &StoredBatch,
+        batch_number: &BatchNumber,
+    ) -> Result<(), anyhow::Error> {
+        self.db.put::<BatchByNumber>(batch_number, batch)?;
+        self.db.put::<BatchByHash>(&batch.hash, batch_number)
+    }
+
+    fn put_transaction(
+        &self,
+        tx: &StoredTransaction,
+        tx_number: &TxNumber,
+    ) -> Result<(), anyhow::Error> {
+        self.db.put::<TxByNumber>(tx_number, tx)?;
+        self.db.put::<TxByHash>(&tx.hash, tx_number)
+    }
+
+    fn put_event(
+        &self,
+        event: &Event,
+        event_number: &EventNumber,
+        tx_number: TxNumber,
+    ) -> Result<(), anyhow::Error> {
+        self.db.put::<EventByNumber>(event_number, event)?;
+        self.db
+            .put::<EventByKey>(&(event.key.clone(), tx_number, event_number.clone()), &())
+    }
+
+    pub fn commit_slot(&self, data_to_commit: SlotCommit) -> Result<(), anyhow::Error> {
+        let item_numbers = {
```

Comment:
Nit: Can we make it a function or add a comment that the scope releases the lock? I am afraid someone might remove the brackets in the future without noticing the lock is released at the end of the scope.

Reply:
Done!